### PR TITLE
Fix #10302 - CalcEquipmentFlowRates assert failure due to out of bounds std::array indexing

### LINKED ISSUE #10302: CalcEquipmentFlowRates assert failure
Issue overview
--------------

An OpenStudio-resources test (`solar_collector_integralcollectorstorage.rb` / osm) that used to run fine is now failing in Debug, and showing All NaNs in the eplustbl.htm in Release.

Here is a backtrace on mac M1, but you'd probably get a more accurate pinpoint on a non M1 GCC (I think it'll throw at the point where the NaN is assigned to this->HotMassFlowRate):

<details>
<summary>**Backtrace: click to expand**</summary>

```
Continuing Simulation at 03/02/2009 for RUN PERIOD 1
Assertion failed: (this->ColdMassFlowRate >= 0.0 && this->ColdMassFlowRate <= this->TotalMassFlowRate), function CalcEquipmentFlowRates, file WaterUse.cc, line 1211.
Process 29040 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = hit program assert
    frame #4: 0x000000010979e2bc libenergyplusapi.24.1.0.dylib`EnergyPlus::WaterUse::WaterEquipmentType::CalcEquipmentFlowRates(this=0x000000013a0ce200, state=0x000000016fdfc7b8) at WaterUse.cc:1211:13
   1208	            this->ColdMassFlowRate = this->TotalMassFlowRate - this->HotMassFlowRate;
   1209	            this->MixedTemp = (this->ColdMassFlowRate * this->ColdTemp + this->HotMassFlowRate * this->HotTemp) / this->TotalMassFlowRate;
   1210	            // there should be no out of bounds results
-> 1211	            assert(this->ColdMassFlowRate >= 0.0 && this->ColdMassFlowRate <= this->TotalMassFlowRate);
   1212	            assert(this->HotMassFlowRate >= 0.0 && this->HotMassFlowRate <= this->TotalMassFlowRate);
   1213	            assert(std::abs(this->HotMassFlowRate + this->ColdMassFlowRate - this->TotalMassFlowRate) < EPSILON);
   1214	        } else {
Target 0: (energyplus) stopped.
(lldb) bt
* thread #1, queue = 'com.apple.main-thread', stop reason = hit program assert
    frame #0: 0x0000000189ef311c libsystem_kernel.dylib`__pthread_kill + 8
    frame #1: 0x0000000189f2acc0 libsystem_pthread.dylib`pthread_kill + 288
    frame #2: 0x0000000189e3aa50 libsystem_c.dylib`abort + 180
    frame #3: 0x0000000189e39d6c libsystem_c.dylib`__assert_rtn + 284
  * frame #4: 0x000000010979e2bc libenergyplusapi.24.1.0.dylib`EnergyPlus::WaterUse::WaterEquipmentType::CalcEquipmentFlowRates(this=0x000000013a0ce200, state=0x000000016fdfc7b8) at WaterUse.cc:1211:13
    frame #5: 0x000000010979ed44 libenergyplusapi.24.1.0.dylib`EnergyPlus::WaterUse::WaterConnectionsType::CalcConnectionsFlowRates(this=0x000000013a0d2a00, state=0x000000016fdfc7b8, FirstHVACIteration=true) at WaterUse.cc:1363:21
    frame #6: 0x000000010979ffac libenergyplusapi.24.1.0.dylib`EnergyPlus::WaterUse::WaterConnectionsType::simulate(this=0x000000013a0d2a00, state=0x000000016fdfc7b8, calledFromLocation=0x000000013976e370, FirstHVACIteration=true, CurLoad=0x000000013976e280, RunFlag=true) at WaterUse.cc:245:19
    frame #7: 0x0000000108ca840c libenergyplusapi.24.1.0.dylib`EnergyPlus::DataPlant::CompData::simulate(this=0x000000013976e200, state=0x000000016fdfc7b8, FirstHVACIteration=true) at Component.cc:113:24
    frame #8: 0x0000000108cb6f38 libenergyplusapi.24.1.0.dylib`EnergyPlus::DataPlant::HalfLoopData::SimulateLoopSideBranchGroup(this=0x000000013a0d0448, state=0x000000016fdfc7b8, FirstBranchNum=2, LastBranchNum=13, FlowRequest=0, FirstHVACIteration=true, LoopShutDownFlag=0x000000016fdf8696) at LoopSide.cc:1763:46
    frame #9: 0x0000000108cb6bbc libenergyplusapi.24.1.0.dylib`EnergyPlus::DataPlant::HalfLoopData::SimulateAllLoopSideBranches(this=0x000000013a0d0448, state=0x000000016fdfc7b8, ThisLoopSideFlow=0, FirstHVACIteration=true, LoopShutDownFlag=0x000000016fdf8696) at LoopSide.cc:614:23
    frame #10: 0x0000000108cb622c libenergyplusapi.24.1.0.dylib`EnergyPlus::DataPlant::HalfLoopData::DoFlowAndLoadSolutionPass(this=0x000000013a0d0448, state=0x000000016fdfc7b8, OtherSide=Supply, ThisSideInletNode=64, FirstHVACIteration=true) at LoopSide.cc:1231:15
    frame #11: 0x0000000108cb558c libenergyplusapi.24.1.0.dylib`EnergyPlus::DataPlant::HalfLoopData::solve(this=0x000000013a0d0448, state=0x000000016fdfc7b8, FirstHVACIteration=true, ReSimOtherSideNeeded=0x000000013a0d0b20) at LoopSide.cc:134:15
    frame #12: 0x0000000108cbce9c libenergyplusapi.24.1.0.dylib`EnergyPlus::PlantManager::ManagePlantLoops(state=0x000000016fdfc7b8, FirstHVACIteration=true, SimAirLoops=0x000000013a01f7c0, SimZoneEquipment=0x000000013a01f7c3, SimNonZoneEquipment=0x000000013a01f7c4, SimPlantLoops=0x000000013a01f7c2, SimElecCircuits=0x000000013a01f7c1) at PlantManager.cc:218:32
    frame #13: 0x00000001084d6d58 libenergyplusapi.24.1.0.dylib`EnergyPlus::HVACManager::SimSelectedEquipment(state=0x000000016fdfc7b8, SimAirLoops=0x000000013a01f7c0, SimZoneEquipment=0x000000013a01f7c3, SimNonZoneEquipment=0x000000013a01f7c4, SimPlantLoops=0x000000013a01f7c2, SimElecCircuits=0x000000013a01f7c1, FirstHVACIteration=0x000000016fdfb3b5, LockPlantFlows=false) at HVACManager.cc:1856:9
    frame #14: 0x00000001084cbd78 libenergyplusapi.24.1.0.dylib`EnergyPlus::HVACManager::SimHVAC(state=0x000000016fdfc7b8) at HVACManager.cc:794:5
    frame #15: 0x00000001084c8b0c libenergyplusapi.24.1.0.dylib`EnergyPlus::HVACManager::ManageHVAC(state=0x000000016fdfc7b8) at HVACManager.cc:344:13
    frame #16: 0x000000010864b5c4 libenergyplusapi.24.1.0.dylib`EnergyPlus::HeatBalanceAirManager::CalcHeatBalanceAir(state=0x000000016fdfc7b8) at HeatBalanceAirManager.cc:4885:9
    frame #17: 0x000000010864b450 libenergyplusapi.24.1.0.dylib`EnergyPlus::HeatBalanceAirManager::ManageAirHeatBalance(state=0x000000016fdfc7b8) at HeatBalanceAirManager.cc:166:5
    frame #18: 0x0000000108766050 libenergyplusapi.24.1.0.dylib`EnergyPlus::HeatBalanceSurfaceManager::ManageSurfaceHeatBalance(state=0x000000016fdfc7b8) at HeatBalanceSurfaceManager.cc:171:5
    frame #19: 0x00000001086eca98 libenergyplusapi.24.1.0.dylib`EnergyPlus::HeatBalanceManager::ManageHeatBalance(state=0x000000016fdfc7b8) at HeatBalanceManager.cc:209:9
    frame #20: 0x0000000109165fb8 libenergyplusapi.24.1.0.dylib`EnergyPlus::SimulationManager::ManageSimulation(state=0x000000016fdfc7b8) at SimulationManager.cc:499:25
    frame #21: 0x0000000107924270 libenergyplusapi.24.1.0.dylib`RunEnergyPlus(state=0x000000016fdfc7b8, filepath="") at EnergyPlusPgm.cc:412:9
    frame #22: 0x0000000107924128 libenergyplusapi.24.1.0.dylib`EnergyPlusPgm(args=size=6, filepath="") at EnergyPlusPgm.cc:240:12
    frame #23: 0x0000000100004f78 energyplus`main at main.cc:67:12
    frame #24: 0x0000000189bb1058 dyld`start + 2224
```
</details>

### Details
Some additional details for this issue (if relevant):
 - Platform (Operating system, version): guessing all, but tested with macOS M1
 - Version of EnergyPlus (if using an intermediate build, include SHA): 23.2.0, develop at 08e876be40e78271d161cdd1a18c40c0b8f737c3

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Defect file added: https://github.com/NREL/EnergyPlusDevSupport/tree/master/DefectFiles/10000s/10302
 - [ ] Ticket added to Pivotal for defect (development team task)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### PR DESCRIPTION
**Pull request by jmarrec**
Pull request overview
---------------------
 - Fix #10302 
 - CalcEquipmentFlowRates assert failure due to out of bounds std::array indexing

### DefectFiles: show fix

cd EnergyPlusDevSupport/DefectFiles/10000s/10302

### before:

image

### After:

image

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [ ] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

File: tst/EnergyPlus/unit/SolarCollectors.unit.cc [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    // Irrelevant to this test
+    constexpr Real64 AirGap = 0.05;  // characteristic length [m]
+    constexpr Real64 CosTilt = 0.87; // cosine of surface tilt angle relative to the horizontal
+    constexpr Real64 SinTilt = 0.80; // sine of surface tilt angle relative to the horizontal
+
+    Real64 hConvCoef = EnergyPlus::SolarCollectors::CollectorData::CalcConvCoeffBetweenPlates(TempSurf1, TempSurf2, AirGap, CosTilt, SinTilt);
+    EXPECT_FALSE(std::isnan(hConvCoef));
+    EXPECT_TRUE(std::isfinite(hConvCoef));
+    EXPECT_NEAR(4.71593, hConvCoef, 0.001);
+}
```
Comment (jmarrec) [inline tst/EnergyPlus/unit/SolarCollectors.unit.cc:78]: Before fix:

```shell
$ Products/energyplus_tests -- --gtest_filter=*CalcConvCoeffBetweenPlates*
Note: Google Test filter = *CalcConvCoeffBetweenPlates*
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from EnergyPlusFixture
[ RUN      ] EnergyPlusFixture.CalcConvCoeffBetweenPlates
/Users/julien/Software/Others/EnergyPlus/tst/EnergyPlus/unit/SolarCollectors.unit.cc:76: Failure
Value of: std::isfinite(hConvCoef)
  Actual: false
Expected: true
/Users/julien/Software/Others/EnergyPlus/tst/EnergyPlus/unit/SolarCollectors.unit.cc:77: Failure
The difference between 4.71593 and hConvCoef is inf, which exceeds 0.001, where
4.71593 evaluates to 4.7159300000000002,
hConvCoef evaluates to -inf, and
0.001 evaluates to 0.001.
[  FAILED  ] EnergyPlusFixture.CalcConvCoeffBetweenPlates (69 ms)
[----------] 1 test from EnergyPlusFixture (69 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (69 ms total)
[  PASSED  ] 0 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] EnergyPlusFixture.CalcConvCoeffBetweenPlates
```

File: src/EnergyPlus/SolarCollectors.hh [hunk 2]
```diff
@@ -54,7 +54,9 @@
 
 // EnergyPlus Headers
 #include <EnergyPlus/DataGlobals.hh>
+#include <EnergyPlus/EPVector.hh>
 #include <EnergyPlus/EnergyPlus.hh>
+#include <EnergyPlus/Plant/DataPlant.hh>
```
Comment (jmarrec) [inline src/EnergyPlus/SolarCollectors.hh:59]: Missing includes

File: src/EnergyPlus/SolarCollectors.cc [hunk 3]
```diff
@@ -1979,7 +1979,7 @@ namespace SolarCollectors {
             CondOfAir = Conductivity[Index];
             PrOfAir = Pr[Index];
             DensOfAir = Density[Index];
-        } else if (Index > NumOfPropDivisions) {
+        } else if (Index >= NumOfPropDivisions) { // 0-index, hence MaxIndex = NumOfPropDivisions - 1
```
Comment (jmarrec) [inline src/EnergyPlus/SolarCollectors.cc:1982]: Actual fix

File: tst/EnergyPlus/unit/SolarCollectors.unit.cc [hunk 4]
```diff
@@ -0,0 +1,78 @@
+// EnergyPlus, Copyright (c) 1996-2023, The Board of Trustees of the University of Illinois,
```
Comment (jmarrec) [inline tst/EnergyPlus/unit/SolarCollectors.unit.cc]:
```suggestion
// EnergyPlus, Copyright (c) 1996-2024, The Board of Trustees of the University of Illinois,
```

Comment (jmarrec) [reply tst/EnergyPlus/unit/SolarCollectors.unit.cc]: I started in 2023 :)

Comment (jmarrec) [reply tst/EnergyPlus/unit/SolarCollectors.unit.cc]: Alright rebased and fixup

**Review by mjwitte — APPROVED** [review]
Fix looks good. Not sure why there aren't any CI results. Updating with develop and will wait for CI to run.

**Comment by Myoldmopar** [comment]
All good here, thanks @jmarrec and @mjwitte, merging.

**Pull request merged**